librz/io/p/io_srec.c

```diff
@@ -0,0 +1,797 @@
+			}
+
+			cksum = bc;
+			cksum += addr_tmp >> 8;
+			cksum += addr_tmp >> 16;
+			cksum += addr_tmp & 0xff;
+
+			if ((next_addr != addr_tmp) || ((sec_size + bc) > SEC_MAX)) {
+				// previous block is not contiguous, or
+				// section buffer is full => write a sparse chunk
+				if (sec_size && sec_size < UT16_MAX) {
+					if (rz_buf_write_at(rbuf, sec_start, sec_tmp, (st32)sec_size) != sec_size) {
+						eprintf("sparse buffer problem, giving up\n");
+						goto fail;
+					}
+				}
+				// advance cursor, reset section
+				sec_start = segreg + addr_tmp;
+				next_addr = addr_tmp;
+				sec_size = 0;
+			}
+			sec_size += bc;
+			next_addr += bc;
+			if (eol) {
+				// checksum
+				if (sscanf(str + 10 + (i * 2), "%02x", &byte) != 1) {
+					eprintf("unparsable data !\n");
+					goto fail;
+				}
+				cksum = ~cksum;
+				if (cksum != byte) {
+					eprintf("Checksum failed (got %02x expected %02x)\n", byte, cksum);
+					goto fail;
+				}
+				*eol = 'S';
+			}
+			str = eol;
+			break;
+
+		case 3: // 32 bit adress, with data field!, kinda same with S1
+
+			// S / 3 / (addres+data+checksum) bytes / address / data / checksum
+			l = sscanf(str + 4, "%08x", &addr_tmp);
+			addr_tmp &= 0xffffffff;
+
+			eol = strchr(str + 1, 'S');
+			if (eol) {
+				*eol = 0;
+			}
+
+			cksum = bc;
+			cksum += addr_tmp >> 8;
+			cksum += addr_tmp >> 16;
+			cksum += addr_tmp >> 24;
+			cksum += addr_tmp & 0xff;
+
+			if ((next_addr != addr_tmp) || ((sec_size + bc) > SEC_MAX)) {
+				// previous block is not contiguous, or
+				// section buffer is full => write a sparse chunk
+				if (sec_size && sec_size < UT16_MAX) {
+					if (rz_buf_write_at(rbuf, sec_start, sec_tmp, (st32)sec_size) != sec_size) {
+						eprintf("sparse buffer problem, giving up\n");
+						goto fail;
+					}
+				}
+				// advance cursor, reset section
+				sec_start = segreg + addr_tmp;
+				next_addr = addr_tmp;
+				sec_size = 0;
+			}
+
+			for (i = 0; i < bc - 5; i++) {
+				if (sscanf(str + 12 + (i * 2), "%02x", &byte) != 1) {
+					eprintf("unparsable data !\n");
+					goto fail;
+				}
+				if (sec_size + i < sec_count) {
+					sec_tmp[sec_size + i] = (ut8)byte & 0xff;
+				}
+				cksum += byte;
+			}
+			sec_size += bc;
+			next_addr += bc;
+			if (eol) {
+				// checksum
+				if (sscanf(str + 12 + (i * 2), "%02x", &byte) != 1) {
+					eprintf("unparsable data !\n");
+					goto fail;
+				}
+				cksum = ~cksum;
+				if (cksum != byte) {
+					eprintf("Checksum failed (got %02x expected %02x)\n", byte, cksum);
+					goto fail;
+				}
+				*eol = 'S';
+			}
+			str = eol;
+			break;
+
+		case 7: // same with S9 but on 32 bits
+
+			// S / 7 / (addres+data+checksum) bytes / address / - / checksum
+			l = sscanf(str + 4, "%08x", &addr_tmp);
+			addr_tmp &= 0xffffffff;
+
+			eol = strchr(str + 1, 'S');
+			if (eol) {
+				*eol = 0;
+			}
+
+			cksum = bc;
+			cksum += addr_tmp >> 8;
+			cksum += addr_tmp >> 16;
+			cksum += addr_tmp >> 24;
+			cksum += addr_tmp & 0xff;
```

Comment:
this checksum seems weird.
if addr_tmp is `0x12345678` then you would do
```c
cksum += 0x56;
cksum += 0x34;
cksum += 0x12;
cksum += 0x78;
```
are you sure about this?
from the doc it should be

```c
cksum += 0x12;
cksum += 0x34;
cksum += 0x56;
cksum += 0x78;
```

Reply:
Well I didn't know the order of the operations matter, the result isn't still the same? because the sum is commutative.

Reply:
that's true.

Reply:
Wait, there might be indeed a problem here, if addr_tmp is `0x12345678`, wouldn't the code result in:
`cksum += 0x00123456
cksum += 0x00001234
cksum += 0x00000012
cksum += 0x78`
Shouldn't there be an `&0xff` added at the end of each sum?